### ISSUE TITLE
lesson-learned: short-circuit within concepts

### ISSUE THREAD

**Issue by marehr** [seqan/product_backlog#286]
### Description 

This is a lesson learned with concepts and an issue that I can point our documentation to.

The following code does not compile with gcc <= 9.x, but does compile with gcc >= 10.0, clang >= 10.0 and msvc v19.23:

```c++
#include <tuple>

template <typename t>
concept always_true = true;

template <typename t>
concept always_false = false;

template <typename t>
concept issue = always_false<t> && always_true<std::tuple_element_t<0, t>>;

static_assert(!issue<int const>); // only compiles for gcc >= 10.0, clang and msvc
static_assert(!issue<int>); // all compilers compile this
```

gcc 9 and below error message:

```
In file included from /opt/compiler-explorer/gcc-9.3.0/include/c++/9.3.0/tuple:38,
                 from <source>:1:
/opt/compiler-explorer/gcc-9.3.0/include/c++/9.3.0/utility: In instantiation of 'struct std::tuple_element<0, const int>':
/opt/compiler-explorer/gcc-9.3.0/include/c++/9.3.0/utility:141:11:   required by substitution of 'template<long unsigned int __i, class _Tp> using tuple_element_t = typename std::tuple_element::type [with long unsigned int __i = 0; _Tp = const int]'
<source>:10:36:   required from here
/opt/compiler-explorer/gcc-9.3.0/include/c++/9.3.0/utility:119:69: error: invalid use of incomplete type 'struct std::tuple_element<0, int>'
  119 |       typedef typename add_const<__tuple_element_t<__i, _Tp>>::type type;
      |                                                                     ^~~~
/opt/compiler-explorer/gcc-9.3.0/include/c++/9.3.0/utility:110:12: note: declaration of 'struct std::tuple_element<0, int>'
  110 |     struct tuple_element;
      |            ^~~~~~~~~~~~~
```

Note that `issue<int const>` will result in `tuple_element_t` being an incomplete type.

```c++
issue<int const> = always_false<int const> && always_true<std::tuple_element_t<0, int const>> (1)
                 = false                   && always_true<typename std::tuple_element<0, int const>::type> (2)
                 = false                   && always_true<typename std::tuple_element<0, int>::type const> (3)
```

where
1. `always_false` is always `false` 
2. `std::tuple_element_t<0, int const>` is defined as `std::tuple_element_t<0, int> const` (the `const` modifier from a tuple will be propagated to the element). See https://en.cppreference.com/w/cpp/utility/tuple/tuple_element.
   ```c++
   template< std::size_t I, class T >
   class tuple_element< I, const T >
   {
     typedef typename std::add_const<typename std::tuple_element<I, T>::type>::type type;
   };
   ```
3. `std::tuple_element_t<0, int>` is incomplete by default, because `int` is not a tuple-like type, and the standard only declares `tuple_element`, but does not define it.

The short answer is, gcc 10 seems to have implemented short-circuiting as described in https://en.cppreference.com/w/cpp/language/constraints#Conjunctions. (I couldn't find the ticket or commit where it was implemented)

-----

I first thought that `std::tuple_element_t<0, t>` is now working for all types `t`, but that is not true.

In general, it is possible to detect incomplete types with concepts, but this only works on its most pure form.

So what do I mean by most pure form?

Let's consider this code:

```c++
template <typename>
struct incomplete;

template <typename t>
struct incomplete<t const> : incomplete<t> {};

template <typename t>
using incomplete_t = typename incomplete<t>::type;
```

See https://godbolt.org/z/daYvcM dor the complete test suite that resulted in this table:

<table>
<thead>
  <tr>
    <th colspan="2" rowspan="3"></th>
    <th colspan="8">always_true = true;<br>always_false = false</th>
    <th colspan="8">always_true = std::is_same&lt;t,t&gt;;<br>always_false = std::is_same&lt;t,void&gt;</th>
  </tr>
  <tr>
    <td colspan="2">gcc&lt;=9</td>
    <td colspan="2">gcc&gt;=10</td>
    <td colspan="2">clang</td>
    <td colspan="2">msvc</td>
    <td colspan="2">gcc&lt;=9</td>
    <td colspan="2">gcc&gt;=10</td>
    <td colspan="2">clang</td>
    <td colspan="2">msvc</td>
  </tr>
  <tr>
    <td>int</td>
    <td>int const</td>
    <td>int</td>
    <td>int const</td>
    <td>int</td>
    <td>int const</td>
    <td>int</td>
    <td>int const</td>
    <td>int</td>
    <td>int const</td>
    <td>int</td>
    <td>int const</td>
    <td>int</td>
    <td>int const</td>
    <td>int</td>
    <td>int const</td>
  </tr>
</thead>
<tbody>
  <tr>
    <td>traits1_incomplete</td>
    <td>always_true&lt;<br>typename<br> incomplete&lt;t&gt;::type&gt;</td>
    <td>false</td>
    <td>hard error</td>
    <td>true</td>
    <td>true</td>
    <td>false</td>
    <td>hard error</td>
    <td>false</td>
    <td>hard error</td>
    <td>false</td>
    <td>hard error</td>
    <td>false</td>
    <td>hard error</td>
    <td>false</td>
    <td>hard error</td>
    <td>false</td>
    <td>hard error</td>
  </tr>
  <tr>
    <td>traits1_incomplete_helper</td>
    <td>always_true&lt;incomplete_t&lt;t&gt;&gt;</td>
    <td>false</td>
    <td>hard error</td>
    <td>true</td>
    <td>true</td>
    <td>false</td>
    <td>hard error</td>
    <td>false</td>
    <td>hard error</td>
    <td>false</td>
    <td>hard error</td>
    <td>false</td>
    <td>hard error</td>
    <td>false</td>
    <td>hard error</td>
    <td>false</td>
    <td>hard error</td>
  </tr>
  <tr>
    <td>traits2_incomplete</td>
    <td>always_false&lt;<br>typename<br> incomplete&lt;t&gt;::type&gt;</td>
    <td>false</td>
    <td>hard error</td>
    <td>false</td>
    <td>false</td>
    <td>false</td>
    <td>hard error</td>
    <td>false</td>
    <td>hard error</td>
    <td>false</td>
    <td>hard error</td>
    <td>false</td>
    <td>hard error</td>
    <td>false</td>
    <td>hard error</td>
    <td>false</td>
    <td>hard error</td>
  </tr>
  <tr>
    <td>traits2_incomplete_helper</td>
    <td>always_false&lt;incomplete_t&lt;t&gt;&gt;</td>
    <td>false</td>
    <td>hard error</td>
    <td>false</td>
    <td>false</td>
    <td>false</td>
    <td>hard error</td>
    <td>false</td>
    <td>hard error</td>
    <td>false</td>
    <td>hard error</td>
    <td>false</td>
    <td>hard error</td>
    <td>false</td>
    <td>hard error</td>
    <td>false</td>
    <td>hard error</td>
  </tr>
  <tr>
    <td>traits3_incomplete</td>
    <td>always_false&lt;t&gt; &amp;&amp; always_true&lt;<br>typename<br> incomplete&lt;t&gt;::type&gt;</td>
    <td>false</td>
    <td>hard error</td>
    <td>false</td>
    <td>false</td>
    <td>false</td>
    <td>false</td>
    <td>false</td>
    <td>false</td>
    <td>false</td>
    <td>hard error</td>
    <td>false</td>
    <td>false</td>
    <td>false</td>
    <td>false</td>
    <td>false</td>
    <td>false</td>
  </tr>
  <tr>
    <td>traits3_incomplete_helper</td>
    <td>always_false&lt;t&gt; &amp;&amp; always_true&lt;incomplete_t&lt;t&gt;&gt;</td>
    <td>false</td>
    <td>hard error</td>
    <td>false</td>
    <td>false</td>
    <td>false</td>
    <td>false</td>
    <td>false</td>
    <td>false</td>
    <td>false</td>
    <td>hard error</td>
    <td>false</td>
    <td>false</td>
    <td>false</td>
    <td>false</td>
    <td>false</td>
    <td>false</td>
  </tr>
  <tr>
    <td>traits4_incomplete</td>
    <td>requires<br>() { <br>typename<br> incomplete&lt;t&gt;::type; } &amp;&amp; always_true&lt;<br>typename<br> incomplete&lt;t&gt;::type&gt;</td>
    <td>hard error</td>
    <td>hard error</td>
    <td>false</td>
    <td>hard error</td>
    <td>false</td>
    <td>hard error</td>
    <td>false</td>
    <td>hard error</td>
    <td>false</td>
    <td>hard error</td>
    <td>false</td>
    <td>hard error</td>
    <td>false</td>
    <td>hard error</td>
    <td>false</td>
    <td>hard error</td>
  </tr>
  <tr>
    <td>traits4_incomplete_helper</td>
    <td>requires<br>() { <br>typename<br> incomplete_t&lt;t&gt;; } &amp;&amp; always_true&lt;incomplete_t&lt;t&gt;&gt;</td>
    <td>hard error</td>
    <td>hard error</td>
    <td>false</td>
    <td>hard error</td>
    <td>false</td>
    <td>hard error</td>
    <td>false</td>
    <td>hard error</td>
    <td>false</td>
    <td>hard error</td>
    <td>false</td>
    <td>hard error</td>
    <td>false</td>
    <td>hard error</td>
    <td>false</td>
    <td>hard error</td>
  </tr>
  <tr>
    <td>traits5_incomplete</td>
    <td>requires<br>() { <br>typename<br> incomplete&lt;std::remove_const_t&lt;t&gt;&gt;::type; } &amp;&amp; always_true&lt;<br>typename<br> incomplete&lt;t&gt;::type&gt;</td>
    <td>false</td>
    <td>hard error</td>
    <td>false</td>
    <td>false</td>
    <td>false</td>
    <td>false</td>
    <td>false</td>
    <td>false</td>
    <td>false</td>
    <td>hard error</td>
    <td>false</td>
    <td>false</td>
    <td>false</td>
    <td>false</td>
    <td>false</td>
    <td>false</td>
  </tr>
  <tr>
    <td>traits5_incomplete_helper</td>
    <td>requires<br>() { <br>typename<br> incomplete_t&lt;std::remove_const_t&lt;t&gt;&gt;; } &amp;&amp; always_true&lt;incomplete_t&lt;t&gt;&gt;</td>
    <td>false</td>
    <td>hard error</td>
    <td>false</td>
    <td>false</td>
    <td>false</td>
    <td>false</td>
    <td>false</td>
    <td>false</td>
    <td>false</td>
    <td>hard error</td>
    <td>false</td>
    <td>false</td>
    <td>false</td>
    <td>false</td>
    <td>false</td>
    <td>false</td>
  </tr>
</tbody>
</table>

(clang version: 12, msvc version: 19.28) 

This table is a bit hard to read, but the gist is that gcc behaves differently from clang and msvc if the concepts `always_true = true` and `always_false = false` definitions don't use the template parameter at all. If gcc does not need to look at the template parameter to evaluate the expression to false/true, it will do so correctly. The other compiler do always return false, if they encounter an incomplete type, even if they don't use them. I don't know yet what the standard behaviour is, but that is a notable difference.

Some other interesting thing to note is that it seems that you can use short-cut definitions like `incomplete_t` and thus `std::tuple_element_t` without any constraint (which was my initial guess what might have been the problem), since they behave exactly as they non-short cut expression.

The (newer) compiler, i.e. gcc>=10, clang and msvc, behave the same, if they don't need to use the type due to a short-circuit (this is shown by traits5_incomplete[_helper]).

And they all fail to use indirect struct definition such as `struct incomplete<t const> : incomplete<t> {};` if the base type is incomplete. This was the main problem for my initial problem with `std::tuple_element`.

The bright side is that in my initial problem this is not so much of a problem, because the (newer) compiler will be able to short-circuit by using the following expression:

```c++
template <typename t>
concept some_tuple_concept = std::tuple_size_v<t> >= 1 && std::same_as<std::tuple_element_t<0, t>, int>;
```

This leaves the question of how to deal with our older compiler support. I guess that this must be decided case-by-case.

For `std::tuple_element_t<0, t>` you could always remove the type-modifier (like `const`, `&`, and so on) to access the direct specialisation of that class `std::tuple_element_t<0, std::remove_cvref_t<t>>`. The main problem is that you will lose the `const` propagation. This might be fine for some use cases, but my use-case needed to distinguish whether the type is "mutable" or not and I thus implemented a work-around where I propagate the `const` myself. The underlying problem wasn't fixed with this, because any custom type could specialise `std::tuple_element` and left it incomplete in some circumstances, but that is a problem that all compiler don't support.

So it seems that it was a correct choice to always make our types complete and thus SFINAE friendly.

----

TLDR;
* gcc>=10 implemented short-circuiting concepts; this makes writing concepts easier
* workarounds for gcc<=9 seem to be always hard in those cases :(
* all our types should always be complete types
